### DISCUSSION TITLE
PlatformIO compilation

### DISCUSSION THREAD

**Discussion by RobertDeHate** [qqqlab/madflight#31] [Q&A] [answered]
I have not had any success compiling in VS code with PlatformIO.
Has anyone done this?
Does anyone have any obvious pointers?
Issues from missing includes to C++ code being compiled as C.
Robert

**Comment by qqqlab** [comment]
I had no problems compiling Quadcopter.ino for ESP32 with PlatformIO: 

- Use "Import Arduino Project" to import Quadcopter.ini
- In Quadcopter.ino change `#include <madflight.h>` to `#include "madflight.h"`
- Add `lib_extra_dirs` to platform.ini (or copy the madflight\src directory to your project's lib directory)

This is my platform.ini:

```
[env:esp32dev]
platform = espressif32
board = esp32dev
framework = arduino
lib_extra_dirs = <enter path to madflight\src here>
```

This is the build output:

```
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (5.2.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
...
Converting 01.Quadcopter.ino
...
Compiling .pio\build\esp32dev\src\01.Quadcopter.ino.cpp.o
...
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [=         ]   5.9% (used 19332 bytes from 327680 bytes)
Flash: [==        ]  24.9% (used 326181 bytes from 1310720 bytes)
Building .pio\build\esp32dev\firmware.bin
esptool.py v4.2.1
Creating esp32 image...
Merged 2 ELF sections
Successfully created esp32 image.
```

**Reply by RobertDeHate** [reply]
Thank you.
This should help many as a starting point.
I will build it just as you did.
Then try the Plane example with my RP2350 and report back.

Robert

**Reply by RobertDeHate** [reply]
OK, so I was able to load the Quadcopter for the ESP32.
So I tried to load the Plane example on the RP2350( which actually shows a RP2040)
I did not find anyway to specify between the two processors and as far as I can tell in the code the RP2040 needs the RTOS but the RP2350 does not. And I have not dug deep enough to see where the FPU of the RP2350 is used as apposed to the RP2040 that lacks it. But maybe that is intrinsic to the compiler that knows?
It seems to hang on this FreeRTOS.
Platform.ini:

```
[env:pico]
platform = raspberrypi
board = pico
framework = arduino
lib_extra_dirs = C:\Firmware\madflight-main

```

Clicked in PIO Explorer window 'src' to see '03.Plane.ino' and clicked to open in editor
Changed #include <madflight.h> to #include "madflight.h"

Build output:
```

C:\Firmware\madflight-main\src/madflight/hw_RP2040/hw_RP2040.h:37:10: fatal error: FreeRTOS.h: No such file or directory

******************************************************************
* Looking for FreeRTOS.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:FreeRTOS.h"
* Web  > https://registry.platformio.org/search?q=header:FreeRTOS.h
*
******************************************************************
```
So I added FreeRTOS-Kernel by linlin as it had support for the RP2040
I got this output:

```In file included from .pio\libdeps\pico\FreeRTOS-Kernel\croutine.c:29:
.pio\libdeps\pico\FreeRTOS-Kernel\include/FreeRTOS.h:59:10: fatal error: FreeRTOSConfig.h: No such file or directory

************************************************************************
* Looking for FreeRTOSConfig.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:FreeRTOSConfig.h"
* Web  > https://registry.platformio.org/search?q=header:FreeRTOSConfig.h
*
************************************************************************
`
```

And can't find a library with "header:FreeRTOSConfig.h"
That is compatible with the RP2350(RP2040)

But plane loaded on Arduino IDE(2.3.3) without issues.
But I just like all the features of VS code(1.95.3) running PIO(Core 6.1.16·Home 3.4.4)

OK one more edit. I had to select 'Raspberry Pi pico' as the board selection as the others would fail.

Robert

**Reply by RobertDeHate** [reply]
To round out the testing I decided to try and compile Quadcopter on the RP pico board and the output was:

```
In file included from C:\Firmware\madflight-main\src/madflight.h:35,
                 from C:/Users/rocke/OneDrive/Documents/PlatformIO/Projects/241201-222006-pico/src/01.Quadcopter.ino:36:
C:\Firmware\madflight-main\src/madflight/hw_RP2040/hw_RP2040.h:37:10: fatal error: FreeRTOS.h: No such file or directory

******************************************************************
* Looking for FreeRTOS.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:FreeRTOS.h"
* Web  > https://registry.platformio.org/search?q=header:FreeRTOS.h
*
******************************************************************

```
And to round it out I loaded the Plane example on the ESP32 with no problem.

I did notice it is particular which ESP32 you select, some fail.
I tried the adafruit feather and it failed and the Pico 2(Raspberry Pi), Pico (Raspberry Pi) failed but Raspberry Pi Pico at least loaded.
But it comes up with two, sometimes 6 different Pico processors at times with the same name and in the boards explorer there seems to be at least two that are different but no way to distinguish them in the import dialog.

Robert

**Reply by qqqlab** [reply]
Yep, Raspberry Pi and PlatformIO are fighting over money, so official support for RP2040/RP2350 is flaky - one more reason to use Arduino IDE/Notepad++ for development ;-)

Have a look at this fork of PlatformIO Raspberry Pi RP2040 which should also support Pico2/RP2350, maybe you get it to work: https://github.com/maxgerhardt/platform-raspberrypi

**Reply by RobertDeHate** [reply]
Thank you for taking a look at it.
Then it wasn't something simple I was missing.
I surly thought the magic fix was going to be, "Change #include <madflight.h> to #include "madflight.h" "

Robert

**Comment by qqqlab** [comment, marked as answer]
Use this platformio.ini file to build madflight for RP2350, RP2040, and ESP32

```ini
; PlatformIO Project Configuration File for madflight

[env]
; Set this to the folder where madflight.h is located
; Default locations when installed via the Arduino IDE:
;    Windows: C:\Users\{username}\Documents\Arduino\libraries\madflight
;    macOS:   /Users/{username}/Documents/Arduino/libraries/madflight
;    Linux:   /home/{username}/Arduino/libraries/madflight
lib_extra_dirs = /ENTER/MADFLIGHT/FOLDER/HERE

; RP2040/RP2350 is not merged into mainline PlatformIO, see 
; https://arduino-pico.readthedocs.io/en/latest/platformio.html
; on Windows you need to enable long path names, see above link for instructions
[env:madflight_pico]
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = pico
framework = arduino
board_build.core = earlephilhower

[env:madflight_pico2]
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = rpipico2
framework = arduino
board_build.core = earlephilhower

[env:madflight_esp32]
platform = espressif32
board = esp32dev
framework = arduino
```

**Reply by RobertDeHate** [reply]
I did finally get it to compile.
I used 02.QuadcopterAdvanced.ino
After I imported it I had to find and rename that file to 02.QuadcopterAdvanced.ino.cpp and add these lines to the beginning
```
#include <Arduino.h>
//========================================================================================================================//
//                                               FUNCTION DECLERATIONS                                                //
//========================================================================================================================//

void i2c_sensors_update();
void imu_loop() ;
void led_Blink() ;
float degreeModulus(float v);
void control_Angle(bool zero_integrators) ;
void control_Rate(bool zero_integrators) ;
void out_KillSwitchAndFailsafe();
void out_Mixer();
```
